### DISCUSSION TITLE
Keep original JsValue during JS->C# interop

### DISCUSSION THREAD

**Discussion by lekrus** [sebastienros/jint#1477] [Q&A]
Hello,

Here is a simple sample class to handle console.log use case:
```c#
var engine = new Engine();
var code = @"
	const obj = {
		a : 'test',
		get b() { console.log('should not invoke'); return 5; }
	}
	console.log(obj);
";
engine.SetValue("console", new JsConsole());
engine.Execute(code);
// var v = engine.GetValue("obj");
Console.ReadKey();

public class JsConsole {
	public void Log(params JsValue[] args) {
		foreach (var arg in args) {
			if(!arg.IsString()) continue;
			Console.WriteLine("Arg: {0}", arg.ToObject());	 
		}
	}
}
```

Is there a way to pass an object ``obj`` into ``JsConsole.Log``, so it's not converted to C# object first? 
My goal is to check all the properties of the passed ``JsValue`` object and skip logging if we have ``FunctionInstance``, but now code in getter ``b`` is always called when passing value into CLR class.

**Comment by SebastianStehle** [comment]
I think you cannot avoid that, but you can probably just introduce a ThreadStatic variable to detect the loop. Like `isLogging`

**Reply by lekrus** [reply]
Thanks for the reply. I'm in fact trying to solve some other use case, but this one with JsConsole is just a good quick sample to show what my question is about. Probably there is some way to get the object without conversion, that is possible, for example, when we do ``engine.GetValue('obj')``

**Comment by SebastianStehle** [comment]
Just loop over the properties manually and do not call ToObject()...

**Reply by SebastianStehle** [reply]
e.g. I have this one: https://github.com/Squidex/squidex/blob/master/backend/src/Squidex.Domain.Apps.Core.Operations/Scripting/ContentWrapper/JsonMapper.cs

I have my own structure for JSON so I have to map it.

**Comment by lahma** [comment]
@lekrus I think this is caused by `CalculateMethodParameterScore` eagerly converting the argument when trying to determine whether the match is good. If you want to work on a PR that would tweak the logic to check already compatible conversions better before calling the `ToObject()` (and with tests to cover) that would be great.

**Reply by lekrus** [reply]
thanks for the hint, will try
